### PR TITLE
layers: l2tp: parse L2TPv2 packets correctly

### PR DESCRIPTION
**Pull request by speidy**
The current code ignores the header flags which makes every L2TP packets
parse as a data message which includes the optional `length`, `nr`, `ns`, `offset` fields.

Signed-off-by: Speidy <speidy@gmail.com>

### CODE REVIEW

**Comment by gpotter2** [comment]
Hi !
Thanks a lot for submitting.

Could you add a regression test ? You will find some examples in test/regression.uts

File: scapy/layers/l2tp.py [hunk 1]
```diff
@@ -16,14 +16,29 @@
 from scapy.layers.inet import UDP
 from scapy.layers.ppp import PPP
 
+header_flags = ['type', 'length', 'res02', 'res03', 'sequence', 'res05',
```
Comment (gpotter2) [inline scapy/layers/l2tp.py]: Please use a more explicit name starting with a _ so that it’s not exported to the entire scapy module

**Comment by speidy** [comment]
@gpotter2 Added a regression test that reflect the changes of parsing.
I also fixed the naming issue as per your comment.

File: scapy/layers/l2tp.py [hunk 2]
```diff
@@ -16,14 +16,29 @@
...
+_fields_present = ['res00', 'res01', 'res02', 'res03', 'priority', 'offset',
+                   'res06', 'sequence', 'res08', 'res09', 'length', 'type']
+
+
 class L2TP(Packet):
-    fields_desc = [ ShortEnumField("pkt_type",2,{2:"data"}),
-                    ShortField("len", None),
+    name = "L2TP"
+    fields_desc = [ 
+                    FlagsField("hdr", 0, len(_fields_present), _fields_present),
```
Comment (p-l-) [inline scapy/layers/l2tp.py]: Could you replace that with something like this (and remove the `_fields_present` global):
```python
    FlagsField("hdr", 0, 12, ['res00', 'res01', 'res02', 'res03', 'priority', 'offset', 'res06', 'sequence',
                                              'res08', 'res09', 'length', 'type']),
```

File: scapy/layers/l2tp.py [hunk 3]
```diff
@@ -16,14 +16,29 @@
...
+                    BitEnumField("version", 2, 4, {2: 'L2TPv2'}),
+
+                    ConditionalField(ShortField("len", 0),
+                        lambda pkt: pkt.hdr.type == 1 or pkt.hdr.length == 1),
```
Comment (p-l-) [inline scapy/layers/l2tp.py]: `lambda pkt: pkt.hdr.type or pkt.hdr.length` or better `lambda pkt: pkt.hdr & 'type+length'`

File: scapy/layers/l2tp.py [hunk 4]
```diff
@@ -16,14 +16,29 @@
...
                     ShortField("tunnel_id", 0),
                     ShortField("session_id", 0),
-                    ShortField("ns", 0),
-                    ShortField("nr", 0),
-                    ShortField("offset", 0) ]
+                    ConditionalField(ShortField("ns", 0),
+                        lambda pkt: pkt.hdr.sequence == 1 or pkt.hdr.type == 1),
```
Comment (p-l-) [inline scapy/layers/l2tp.py]: Same here.

File: scapy/layers/l2tp.py [hunk 5]
```diff
@@ -16,14 +16,29 @@
...
+                    ConditionalField(ShortField("nr", 0),
+                        lambda pkt: pkt.hdr.sequence == 1 or pkt.hdr.type == 1),
```
Comment (p-l-) [inline scapy/layers/l2tp.py]: Same here.

File: scapy/layers/l2tp.py [hunk 6]
```diff
@@ -16,14 +16,29 @@
...
+                    ConditionalField(
+                        PadField(ShortField("offset", 0), 4, b"\x00"),
+                        lambda pkt: pkt.hdr.type == 0 and pkt.hdr.offset == 1
```
Comment (p-l-) [inline scapy/layers/l2tp.py]: Same here.

**Comment by speidy** [comment]
@p-l- Thanks for reviewing.
Iv'e pushed a fix for your remarks.

Also, a real-life example for incorrectly parsed L2TP packets can be tested using this [pcap file](https://sourceforge.net/p/jnetstream/svn/332/tree/resources/captures/l2tp.pcap). As you can see the offset and sequence bits are off for this packets; the current code ignores the flags and incorrectly parse data as those fields.

**Comment by p-l-** [comment]
Thanks!

**Pull request merged**